### DISCUSSION TITLE
LND architectural overview

### DISCUSSION THREAD

**Discussion by trevlt** [lightningnetwork/lnd#7896] [Q&A]
Is there a document that gives an architectural overview of the LND code base?

**Comment by guggero** [comment]
Unfortunately there is none, at least no up-to-date one that I am aware of.
Depending on what your goal is, diving into the code base might be easiest if you start at an RPC endpoint. For example, if you want to know where the channel list is stored, you'd go to the [`ListChannels` RPC in `rpcserver.go`](https://github.com/lightningnetwork/lnd/blob/ab117a0ed444b920a574ec0bbc002fb0247bf6a6/rpcserver.go#L4059) and then dive into the components invoked from there. It might be a good idea to use an IDE that allows you to navigate the code by just clicking on methods/functions/types.

If you want to know more about what components are started and how they are dependent on each other, then [`server.go`](https://github.com/lightningnetwork/lnd/blob/1a98fcff7e5543e681440c90922db7339dd18da4/server.go#L472) might be a good start. That initializes all the main sub-components in order of their dependencies.

**Comment by trevlt** [comment]
Thanks @guggero. I'm wondering what is a rational to have all the packages in the root directory and not grouped in sub-directories per common context. It gets very hard to read. Is it some new idiomatic way to write a Go project, or is it something specific to LND?

**Reply by guggero** [reply]
Yes, that's true, the list has gotten very long. I guess it's just because it grew organically. And not having a very deep directory structure helps to find things more easily in some situation. Maybe things could be grouped just a bit more. But then again, refactor PRs just for the sake of refactor aren't the most popular to review.
